src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableExtensions.cs

```diff
@@ -50,10 +50,12 @@ internal static IReadOnlyList<T> AsReadOnlyList<T>(this IEnumerable<T> sequence)
                 return orderedCollection;
             }
 
+#if !NET10_0_OR_GREATER // ICollection<T> : IReadOnlyCollection<T> on .NET 10+
             if (sequence is IList<T> listOfT)
```

Comment:
Small question: technically speaking the DIMs changes in #115802 changed the semantics here, right? Suppose I had `MyMutableList<int>` that only implemented `IList<int>`. If I called `AsReadOnlyList` before #115802, that would take this path, returning an immutable wrapper, so trying to downcast wouldn't work. Whereas with DIMs (regardless of this PR), the path above would be taken, returning the same object, meaning someone could then just downcast and mutate the object. I assume that's by design and downcasting in general is bad either way? 🤔